lightningd/peer_control.c

```diff
@@ -324,6 +335,72 @@ register_close_command(struct lightningd *ld,
 			     &close_command_timeout, cc);
 }
 
+/* Destroy the open command structure in reaction to the
+ * channel being destroyed. */
+static void
+destroy_open_command_on_channel_destroy(struct channel *_ UNUSED,
+					struct open_command *oc)
+{
+	/* The oc has the command as parent, so resolving the
+	 * command destroys the oc and triggers destroy_open_command.
+	 * Clear the oc->channel first so that we will not try to
+	 * remove a destructor. */
+	oc->channel = NULL;
+	was_pending(command_fail(oc->cmd, LIGHTNINGD,
+				 "Channel forgotten before open concluded."));
+}
+
+/* Destroy the open command structure. */
+static void
+destroy_open_command(struct open_command *oc)
+{
+	list_del(&oc->list);
+	/* If destroy_close_command_on_channel_destroy was
+	 * triggered beforehand, it will have cleared
+	 * the channel field, preventing us from removing it
+	 * from an already-destroyed channel. */
+	if (!oc->channel)
+		return;
+	tal_del_destructor2(oc->channel,
+			    &destroy_open_command_on_channel_destroy,
+			    oc);
+}
+
+/* Handle timeout. */
+static void
+open_command_timeout(struct open_command *oc)
+{
+	/* Fail the command directly, which will resolve the
+	 * command and destroy the open_command. */
+	was_pending(command_fail(oc->cmd, LIGHTNINGD,
+				 "Channel open not finished "
+				 "before timeout"));
+}
+
+
+void register_open_command(struct lightningd *ld,
+			   struct command *cmd,
+			   struct channel *channel,
+			   unsigned int *timeout)
+{
+	struct open_command *oc;
+	assert(channel);
+
+	oc = tal(cmd, struct open_command);
+	list_add_tail(&ld->open_commands, &oc->list);
+	oc->cmd = cmd;
+	oc->channel = channel;
+	tal_add_destructor(oc, &destroy_open_command);
+	tal_add_destructor2(channel,
+			    &destroy_open_command_on_channel_destroy,
+			    oc);
```

Comment:
No, this is the wrong level!  Just remove the timeout for now.

There should be a timeout on the *channel construction* which does this, and more; actually broadcasts the unilateral or simply forgets (if we have no inputs).  This is a general fix which solves #4043 too!